File: Session3_DP/unique-path.java
class Solution {
    public int uniquePaths(int m, int n) {
        int[][] dp = new int[n][m];
        /* Only 1 unique path if walking straight to the right */
        for (int i = 0; i < n; i++) {
            dp[i][0] = 1;
        }
        
        /* Only 1 unique path if walking straight downwards */
        for (int j = 0; j < m; j++) {
            dp[0][j] = 1;
        }
        
        /**
         * CRITICAL: Here, the DP Formula is basically the sum
         * of number of ways of reaching square above and square
         * to the left of the current square.
         */
        for (int i = 1; i < n; i++) {
            for (int j = 1; j < m; j++) {
                dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
            }   
        }
        
        return dp[n - 1][m - 1];
    }
}

File: Session3_DP/unique-path.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        # Set 1 Unique path (Since what's important is
        # Top and Left border, with only 1 unique path
        # Since it is straight.
        dp = [[1 for _ in range(n)] for _ in range(m)]

        # CRITICAL: Here, the DP Formula is basically the sum
        # of number of ways of reaching square above and square
        # to the left of the current square.
        for i in range(1, n):
            for j in range(1, m):
                dp[j][i] = dp[j - 1][i] + dp[j][i - 1]
        return dp[m - 1][n - 1]